File: src/jpa.js
// JPA / Hibernate pass: entities + tables, relationships (cascade/fetch/mappedBy/
// owning/join), and Spring Data repositories linked to the entity they manage.
import * as db from './db.js';

const RELATION_ANNS = {
  OneToMany: 'one_to_many', ManyToOne: 'many_to_one',
  OneToOne: 'one_to_one', ManyToMany: 'many_to_many',
};
const COLLECTION_RELATIONS = new Set(['one_to_many', 'many_to_many']);
const DEFAULT_FETCH = {
  one_to_many: 'LAZY', many_to_many: 'LAZY', many_to_one: 'EAGER', one_to_one: 'EAGER',
};
const REPO_BASES = new Set([
  'JpaRepository', 'CrudRepository', 'PagingAndSortingRepository',
  'ReactiveCrudRepository', 'Repository', 'JpaSpecificationExecutor',
]);

const simple = (n) => (n.includes('.') ? n.split('.').pop() : n);
const strArg = (args, key) => { const m = args.match(new RegExp(`${key}\\s*=\\s*"([^"]*)"`)); return m ? m[1] : null; };
const boolArg = (args, key) => { const m = args.match(new RegExp(`${key}\\s*=\\s*(true|false)`)); return m ? m[1] === 'true' : null; };
const intArg = (args, key) => { const m = args.match(new RegExp(`${key}\\s*=\\s*(\\d+)`)); return m ? Number(m[1]) : null; };
const enumTokens = (args, en) => [...args.matchAll(new RegExp(`${en}\\.(\\w+)`, 'g'))].map((m) => m[1]);
const genericFirst = (s) => { const m = s.match(/<\s*([A-Za-z_][A-Za-z0-9_.]*)/); return m ? simple(m[1]) : null; };

function tableName(ci) {
  const t = ci.annotation('Table');
  if (t && t.argsText) return strArg(t.argsText, 'name') || (t.argsText.match(/"([^"]+)"/)?.[1]) || ci.name;
  return ci.name;
}

export function isEntity(ci) { return ci.hasAnnotation('Entity'); }

export function repositoryEntity(ci) {
  for (const raw of ci.rawSupertypes) {
    const base = simple(raw.split('<')[0]).trim();
    if (REPO_BASES.has(base) && raw.includes('<')) { const e = genericFirst(raw); if (e) return e; }
  }
  return null;
}

function columnMapping(f) {
  const names = new Set(f.annotations.map((a) => a.name));
  const col = f.annotation('Column');
  const join = f.annotation('JoinColumn');
  const gen = f.annotation('GeneratedValue');
  let columnName = col && col.argsText ? strArg(col.argsText, 'name') : null;
  if (!columnName && join && join.argsText) columnName = strArg(join.argsText, 'name');
  const m = {
    field: f.name, type: f.typeSimple, column: columnName || f.name,
    primary_key: names.has('Id') || names.has('EmbeddedId'),
  };
  if (gen) { const s = enumTokens(gen.argsText || '', 'GenerationType'); m.generated = s[0] || 'AUTO'; }
  if (col && col.argsText) {
    const n = boolArg(col.argsText, 'nullable'); const u = boolArg(col.argsText, 'unique'); const l = intArg(col.argsText, 'length');
    if (n !== null) m.nullable = n; if (u !== null) m.unique = u; if (l !== null) m.length = l;
  }
  for (const [flag, ann] of [['enumerated', 'Enumerated'], ['lob', 'Lob'], ['version', 'Version'], ['embedded', 'Embedded'], ['transient', 'Transient']]) {
    if (names.has(ann)) m[flag] = true;
  }
  return m;
}

function relationAttrs(f, rel) {
  const relAnn = f.annotations.find((a) => RELATION_ANNS[a.name]);
  const args = relAnn ? relAnn.argsText : '';
  const mappedBy = strArg(args, 'mappedBy');
  const fetchExplicit = enumTokens(args, 'FetchType');
  const cascade = enumTokens(args, 'CascadeType');
  const orphan = boolArg(args, 'orphanRemoval');
  const attrs = {
    field: f.name, fetch: fetchExplicit[0] || DEFAULT_FETCH[rel], fetch_default: fetchExplicit.length === 0,
    owning: mappedBy === null,
  };
  if (mappedBy) attrs.mapped_by = mappedBy;
  if (cascade.length) attrs.cascade = cascade;
  if (orphan !== null) attrs.orphan_removal = orphan;
  const join = f.annotation('JoinColumn');
  if (join) { attrs.join_column = strArg(join.argsText || '', 'name') || f.name; const r = strArg(join.argsText || '', 'referencedColumnName'); if (r) attrs.referenced_column = r; }
  const jt = f.annotation('JoinTable');
  if (jt) attrs.join_table = strArg(jt.argsText || '', 'name');
  return attrs;
}

export function apply(database, classes, symbols) {
  for (const ci of classes) {
    if (isEntity(ci)) {
      const columns = ci.fields.map(columnMapping);
      const attrs = { table: tableName(ci), columns };
      database.prepare('UPDATE nodes SET layer=?, signature=?, attrs=? WHERE id=?')
        .run('entity', `table=${attrs.table}`, JSON.stringify(attrs), ci.id);
      ci.fields.forEach((f, i) => db.setNodeAttrs(database, `${ci.id}.${f.name}`, JSON.stringify(columns[i])));
      for (const f of ci.fields) {
        const relName = f.annotations.find((a) => RELATION_ANNS[a.name]);
        if (!relName) continue;
        const rel = RELATION_ANNS[relName.name];
        let targetSimple = f.typeSimple;
        if (COLLECTION_RELATIONS.has(rel) && f.typeRaw && f.typeRaw.includes('<')) targetSimple = genericFirst(f.typeRaw) || targetSimple;
        const t = symbols.resolve(targetSimple, ci);
        if (t && t !== ci.id) db.addEdge(database, ci.id, t, rel, JSON.stringify(relationAttrs(f, rel)));
      }
    }
    const entitySimple = repositoryEntity(ci);
    if (entitySimple) {
      database.prepare("UPDATE nodes SET layer='repository' WHERE id=?").run(ci.id);
      const t = symbols.resolve(entitySimple, ci);
      if (t) db.addEdge(database, ci.id, t, 'persists');
    }
  }
}

File: code_kg/jpa.py
"""JPA / Hibernate awareness pass (deterministic).

Adds the persistence layer on top of the generic spine + Spring pass:
  * Entities    — @Entity / @Table classes -> layer "entity"; table name captured
                  in the node's signature column as "table=<name>".
  * Relations   — @OneToMany / @ManyToOne / @OneToOne / @ManyToMany fields ->
                  edges between entity classes (kind = one_to_many / many_to_one /
                  one_to_one / many_to_many). Collection element type is resolved
                  from the generic argument (List<Order> -> Order).
  * Repositories — interfaces extending Spring Data's JpaRepository<T, ID> (and
                  friends) -> layer "repository" + a `persists` edge to entity T.

Consumes the ``ClassInfo`` model from ``extract.py`` (no re-parsing). Runs after
the Spring pass so it can refine layers for entities and JPA repositories.
"""

from __future__ import annotations

import re
from typing import Optional

from .extract import ClassInfo, Symbols

RELATION_ANNS = {
    "OneToMany": "one_to_many",
    "ManyToOne": "many_to_one",
    "OneToOne": "one_to_one",
    "ManyToMany": "many_to_many",
}

COLLECTION_RELATIONS = {"one_to_many", "many_to_many"}

# Spring Data repository base interfaces; first generic arg is the managed entity.
REPO_BASES = {
    "JpaRepository",
    "CrudRepository",
    "PagingAndSortingRepository",
    "ReactiveCrudRepository",
    "Repository",
    "JpaSpecificationExecutor",
}

_GENERIC_FIRST = re.compile(r"<\s*([A-Za-z_][A-Za-z0-9_.]*)")
_GENERIC_ALL = re.compile(r"<\s*([A-Za-z_][A-Za-z0-9_.]*)")


def _simple(name: str) -> str:
    return name.rsplit(".", 1)[-1] if "." in name else name


def _table_name(ci: ClassInfo) -> str:
    """Resolve the table name: @Table(name=...) > @Table("...") > class name."""
    t = ci.annotation("Table")
    if t and t.args_text:
        m = re.search(r'name\s*=\s*"([^"]+)"', t.args_text) or re.search(r'"([^"]+)"', t.args_text)
        if m:
            return m.group(1)
    return ci.name


def is_entity(ci: ClassInfo) -> bool:
    return ci.has_annotation("Entity")


def repository_entity(ci: ClassInfo) -> Optional[str]:
    """If ``ci`` is a Spring Data repository, return the managed entity simple name."""
    for raw in ci.raw_supertypes:
        base = _simple(raw.split("<", 1)[0]).strip()
        if base in REPO_BASES and "<" in raw:
            m = _GENERIC_FIRST.search(raw)
            if m:
                return _simple(m.group(1))
    return None


def apply(conn, classes: list[ClassInfo], symbols: Symbols) -> None:
    from . import db

    for ci in classes:
        # entities
        if is_entity(ci):
            conn.execute(
                "UPDATE nodes SET layer = 'entity', signature = ? WHERE id = ?",
                (f"table={_table_name(ci)}", ci.id),
            )
            _apply_relations(conn, ci, symbols)

        # spring data repositories
        entity_simple = repository_entity(ci)
        if entity_simple:
            conn.execute("UPDATE nodes SET layer = 'repository' WHERE id = ?", (ci.id,))
            tgt = symbols.resolve(entity_simple, ci)
            if tgt:
                db.add_edge(conn, ci.id, tgt, "persists")

    conn.commit()


def _apply_relations(conn, ci: ClassInfo, symbols: Symbols) -> None:
    from . import db

    for fld in ci.fields:
        rel = None
        for ann in fld.annotations:
            if ann.name in RELATION_ANNS:
                rel = RELATION_ANNS[ann.name]
                break
        if rel is None:
            continue

        # resolve the target entity simple name
        target_simple = fld.type_simple
        if rel in COLLECTION_RELATIONS and fld.type_raw and "<" in fld.type_raw:
            m = _GENERIC_FIRST.search(fld.type_raw)
            if m:
                target_simple = _simple(m.group(1))

        tgt = symbols.resolve(target_simple, ci)
        if tgt and tgt != ci.id:
            db.add_edge(conn, ci.id, tgt, rel)
